src/main/java/com/xjeffrose/xio/http/Http2ClientCodec.java

```diff
@@ -0,0 +1,137 @@
+package com.xjeffrose.xio.http;
+
+import com.xjeffrose.xio.core.internal.UnstableApi;
+import com.xjeffrose.xio.http.internal.FullHttp2Response;
+import com.xjeffrose.xio.http.internal.Http2StreamingData;
+import com.xjeffrose.xio.http.internal.StreamingHttp2Response;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http2.DefaultHttp2DataFrame;
+import io.netty.handler.codec.http2.DefaultHttp2Headers;
+import io.netty.handler.codec.http2.Http2DataFrame;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.util.AttributeKey;
+import io.netty.util.concurrent.PromiseCombiner;
+import lombok.extern.slf4j.Slf4j;
+
+@UnstableApi
+@Slf4j
+public class Http2ClientCodec extends ChannelDuplexHandler {
+
+  private static final AttributeKey<Response> CHANNEL_RESPONSE_KEY =
+      AttributeKey.newInstance("xio_channel_h2_response");
+
+  private static void setChannelResponse(ChannelHandlerContext ctx, Response response) {
+    ctx.channel().attr(CHANNEL_RESPONSE_KEY).set(response);
+  }
+
+  private static Response getChannelResponse(ChannelHandlerContext ctx) {
+    // TODO(CK): Deal with null?
+    return ctx.channel().attr(CHANNEL_RESPONSE_KEY).get();
+  }
+
+  Response wrapHeaders(Http2Headers headers, int streamId, boolean eos) {
+    if (eos) {
+      return new FullHttp2Response(headers, streamId);
+    } else {
+      return new StreamingHttp2Response(headers, streamId);
+    }
+  }
+
+  Response wrapResponse(ChannelHandlerContext ctx, Http2Response msg) {
+    log.debug("wrapResponse msg={}", msg);
+    if (msg.payload instanceof Http2Headers) {
+      Http2Headers headers = (Http2Headers) msg.payload;
+      if (msg.eos && headers.method() == null && headers.status() == null) {
+        Response response =
+            new StreamingResponseData(getChannelResponse(ctx), new Http2StreamingData(headers));
+        return response;
+      } else {
+        Response response = wrapHeaders(headers, msg.streamId, msg.eos);
+        setChannelResponse(ctx, response);
+        return response;
+      }
+    } else if (msg.payload instanceof Http2DataFrame) {
+      Response response =
+          new StreamingResponseData(
+              getChannelResponse(ctx),
+              new Http2StreamingData(((Http2DataFrame) msg.payload).content(), msg.eos));
+      return response;
+    }
+    // TODO(CK): throw an exception?
+    return null;
```

Comment:
If we throw in the pipeline, we need to do a better job of doing something more meaningful with the exception handler

Reply:
Throwing in the pipeline pretty much means we're fucked and the connection is going to be closed. I'm okay with that for situations like this where we've reached a bad state and things just need to die.